src/api/MainSolver.cc

```diff
@@ -135,10 +135,12 @@ sstat MainSolver::simplifyFormulas() {
                 PTRef processed = theory->preprocessAfterSubstitutions(fla, context);
                 pmanager.transferPartitionMembership(fla, processed);
                 frameFormulas.push(processed);
+                preprocessor.addProcessedFormula(processed);
             }
             if (frameFormulas.size() == 0 or std::all_of(frameFormulas.begin(), frameFormulas.end(), [&](PTRef fla) { return fla == logic.getTerm_true(); })) {
                 continue;
             }
+            theory->formulaAfterPreprocessing(preprocessor.getPreprocessedFormulas());
```

Comment:
It always requires `preprocessor.getPreprocessedFormulas()` as an argument so it would be suitable to default the argument, but `preprocessor` is private and not accessible from `Theory`. `Preprocessor` is probably not meant to be public, for example `getTheory` is also non-public. But then it is not possible to make `Theory` work directly with `preprocessor`. Friendship is not a solution here, it does not propagate to inherited classes. I must admit that I am quite confused at the moment what is the purpose of `Theory` since it seems to serve only for preprocessing reasons. Then it is weird that we have both `Theory` and `Preprocessor`, not connected with each other ... I do not require to fix this right now, but at least some brief documentation and/or mentioning how to fix it in the future would help.

Reply:
Ideally, I would eliminate `Theory`, but preprocessing is thery-specific, so the question is how should that be set up.

Reply:
I think that the idea of the `Theory` class hierarchy is OK, but I would place it inside of the class `Preprocess`. So then it can be `UFPreprocess` etc. This would just require to use a pointer on the `preprocess` object.

Reply:
I am leaning in this direction as well.
Make `Preprocessor` a standalone class (outside of `MainSolver`) and have subclasses (or other customization) for different theories.
But there are still a few points that need to be resolved.

Reply:
Not convinced about `Preprocessor` being standalone (outside), but overall I agree.